generator/internal/parser/httprule/http_rule_parser.go

```diff
@@ -0,0 +1,191 @@
+// Reference:
+// - https://go.dev/talks/2011/lex.slide (slides)
+// - https://go.dev/talks/2011/lex/r59-lex.go (code)
+
+func Parse(pathTemplate string) ([]api.PathSegment, error) {
+
+	l := &lexer{
+		input:    pathTemplate,
+		state:    slashState, // the first state is always slashState
+		segments: make(chan segment, 2),
+	}
+	var segments []api.PathSegment
+	for {
+		s := l.nextItem()
+		switch s.typ {
+		case segmentLiteral:
+			segments = append(segments, api.NewLiteralPathSegment(s.val))
+		case segmentIdentifier:
+			segments = append(segments, api.NewFieldPathPathSegment(s.val))
+		case segmentVerb:
+			segments = append(segments, api.NewVerbPathSegment(s.val))
+		case segmentError:
+			return nil, fmt.Errorf("error parsing path template (%s): %s", pathTemplate, s.val)
+		case segmentEOF:
+			return segments, nil
+		}
+
+	}
+}
+
+// ### Path template syntax
+//
+//	Template = "/" Segments [ Verb ] ;
+//	Segments = Segment { "/" Segment } ;
+//	Segment  = "*" | "**" | LITERAL | Variable ;
+//	Variable = "{" FieldPath [ "=" Segments ] "}" ;
+//	FieldPath = IDENT { "." IDENT } ;
+//	Verb     = ":" LITERAL ;
+const (
+	segmentError segmentType = iota // 0
+	segmentLiteral
+	segmentIdentifier
+	segmentVerb //a verb is actually ':' + LITERAL,
+	// but we need a way to differentiate a literal within a segment, from a literal within a verb
+	segmentEOF
+)
+
+const (
+	slash    = '/'
+	star     = '*'
+	varLeft  = '{'
+	varRight = '}'
+	varSep   = '='
+	verbSep  = ':'
+)
+
+type stateFn func(*lexer) stateFn
+
+// slashState is the first state before a segment
+func slashState(l *lexer) stateFn {
+	if !l.ignoreIfMatches(slash) {
+		return l.unexpectedRuneError(slash, l.peek())
+	}
+	return segmentState
+}
+
+func segmentState(l *lexer) stateFn {
+	switch l.peek() {
+	case eof, slash, verbSep:
+		return l.errorf("expected a segment, found %q", l.peek())
+	case varLeft:
+		return varState
+	}
+	return literalState
+}
+
+const validFirstIdentifierRunes = "ABCDEFGHIJKLMNOPQRSTUVWXYZ" +
+	"abcdefghiljkmnopqrstuvwxyz"
+
+const validIdentifierRunes = validFirstIdentifierRunes +
+	"0123456789" +
+	"_"
+const validVariableValueRunes = validIdentifierRunes + "/*"
+
+func varState(l *lexer) stateFn {
+	if !l.ignoreIfMatches(varLeft) {
+		return l.unexpectedRuneError(varLeft, l.peek())
+	}
+
+	if !l.accept(validFirstIdentifierRunes) {
+		return l.errorf("expected a valid first rune for a variable identifier, got %q", l.peek())
+	}
+
+	// The identifier is a single or multiple characters, just take all that match.
+	_ = l.acceptAll(validIdentifierRunes)
+
+	// We have reached the end of the valid characters for an identifier
+	// emit the segment now and process the rest of the variable for correctness.
+	l.emit(segmentIdentifier)
+
+	if l.ignoreIfMatches(varSep) {
+		// This var has a value, we don't use it for anything, so lets just ignore it.
+		//TODO The valid format for a variable value is a sequence of segments, we could validate that here.
```

Comment:
Do we want to validate this? I couldn't find a way to contribute the value back into the expected data structure, so I didn't bother parsing it.